#5: Added design document (#12)
exasol/advanced-analytics-framework: doc/design.md

# Design

## Introduction

> This library is an event-driven framework that allows users to build complex 

Reviewer: We should introduce the framework more generally and then maybe state that we choose a event driven design because of the requirements and the limitation that pquery only is available for Lua at the moment.

PS: A dataflow based programming model/interface for the user is still possible. I will do something similar for machine learning. Where I build an execution graph and have an event driven graph runner which visits in the event handler the next node in the graph und returns its queries. I am only lazy regarding loops, which handle by calling another instance of this framework. However, this should be possible to incorporate as well. We might add this dataflow based programming model to this library, but not now.

- The two main components of the framework, which are an Event Loop and an Event 
Handler, should be decoupled so that they can be independently updated and 
implemented. It also allows us to replace them with a different implementation 
in another programming language.

- The Event Handler should be able to keep track of the status of a particular 
operation by storing states in BucketFS.

## System Design and Architecture
 
TBD 

## Event Loop
Event Loop  processes only the state transitions by executing queries returned 
by the Event Handler. This loop provides the new state transition by calling 
the Event Handler until the Event Handler returns the stop signal with a result 
or an error.

### Initiating Event Loop
`dsn~initiating-event-loop~1`

The Event Loop is responsible for getting the parameters of the Event Handler 
that will be called in the loop and forward them to it.

Covers:

* `req~intiating-global-loop~1`

Needs: impl, utest, itest

### Handling States
`dsn~handling-states~1`

The  output of the called Event Handler query can include SQL queries to 
complete its computation. In that case Event Loop run them on each iteration, 

> until the Event Handler returns the status as complete.

Reviewer: The headline and the description don't fit, and later there is another section with storing the state in bucketfs. We only need to handle and store the state in the framework, because the execution of the event handler happens in a UDF and the UDF ends after each call to the event handler.

`dsn~managing-temporary-bucketfs-files~1`

The Event Handler create temporary BucketFS files that will be  kept as result. 
These temporary files are placed  in the same directory in  the BucketFS.   

Covers:

* `req~managing-temporary-bucketfs-files~1`

Needs: impl, utest, itest

### Managing Temporary BucketFS Files
`dsn~managing-temporary-tables~1`

The Event Handler uses temporary tables to store large intermediate results. 

Covers:

* `req~managing-temporary-tables~1`

Needs: impl, utest, itest

### Error Handling
In case of any errors in the execution, temporarily created files and tables 
are cleaned. 

#### Cleanup Temporary Tables
`dsn~cleanup-temporary-tables~1`

Temporary tables are removed in case of an error that occur during the execution 
of the framework with the keeping option  not selected.

Covers:

* `req~cleanup-temporary-tables~1`

Needs: impl, utest, itestn

#### Cleanup Temporary BucketFS Files
`dsn~cleanup-temporary-bucketfs-files~1`

Temporary BucketFS files are removed in case of an error that occur during 
the execution of the framework with the keeping option  not selected.

Covers:

* `req~cleanup-temporary-bucketfs-files~1`

> Needs: impl, utest, itest

Author: preparing a new language container could be challenge for some users. I'm in favor of creating python udf script.

> Needs: impl, utest, itest

Reviewer: we need both here, for the machine learning library we will build a separate language container

The framework keeps states during iterations by storing them in BucketFS.

Covers:

* `req~managing-temporary-bucketfs-files~1`

Needs: impl, utest, itest

It returns the sequence of SQL queries to execute
A query to call the same UDF to run the action for the next state

### Returning Queries
`dsn~returning-queries~1`

The Event Handler returns a list of SQL queries to execute and the return query 
which will be called again as the next state action so that the new  state 
transition is performed.

Covers:

* `feat~implementation-framework~1`

Needs: impl, utest, itest

### Managing Temporary BucketFS Files
`dsn~managing-temporary-bucketfs-files~1`

The Event Handler create temporary BucketFS files that will be  kept as result. 
These temporary files are placed  in the same directory in  the BucketFS.   

Covers:

* `req~managing-temporary-bucketfs-files~1`

Needs: impl, utest, itest

### Managing Temporary BucketFS Files
`dsn~managing-temporary-tables~1`

The Event Handler uses temporary tables to store large intermediate results. 

Covers:

* `req~managing-temporary-tables~1`

Needs: impl, utest, itest

### Electing Leader
`dsn~electing-leader~1`

> 

Reviewer: Before leader election can happen. UDFs need to discover each other. We will use pycos, but it works, in general, the following. Every UDF receives on a fixed port UDP packages and sends UDP broadcasts on the same port. This way, everyone should find everyone. Assuming the Network has enough capacity left to send the UDP packages. To connect not to UDF instances of a different UDF call, each UDF call includes a unique ID which will be part of the broadcast message. We use this ID to connect only to other UDF instances of the same UDF Call.

Further, we know how many UDF instances should have been started, and we wait with leader election, until all instances have connected to each other.

# Design

## Introduction
This document describes the design of the advanced analytics framework on which 
complex data analysis algorithms can be built.

## Constraints / Limitations

- The best way to run dynamic SQL queries is to use Lua scripts. Because Lua
scripting enables us to run queries in the same transaction using the `pquery` method.
- The results of `pquery` are simple handles and there is a limit of 250 open 
result set handles for Lua scripts.
- The main logic of the framework should be implemented in Python, because it is 
widely used in data processing and analysis.
- Python UDF, where the main logic is placed, returns the result of the query 
and finishes its execution. Due to that, it cannot be kept running and used 
as server.

## Assumptions

- Since the number open result set handles in Lua scripts is limited, it is 
assumed that Lua scripts is not nested more than 250 times.
- The result of `pquery` is assumed to not exceed the db memory limit in size. 
The DB usually allows 4 GB per query and node. The UDF can consume up to 32 GB 
per node, but gets throttled down after the 4 GB. However, the 32 GB are shared 
between all queries on a node.

> - The state of the algorithm is not too large.

Reviewer: ```suggestion
- Because Python UDF can't run constantly, we need to save its state to the BucketFS. Saving and loading needs to reasonable fast to not introduce too much overhead. For that, we assume the state of the Python UDF is reasonable small (max a few MB).
```

the Event Handler until the Event Handler returns the stop signal with a result 
or an error.

### Initiating Event Loop
`dsn~initiating-event-loop~1`

The Event Loop is responsible for getting the parameters of the Event Handler 
that will be called in the loop and forward them to it.

Covers:

* `req~intiating-algorithm~1`

Needs: impl, utest, itest

### Iterating over Loop
`dsn~handling-states~1`

The  output of the called Event Handler query can include SQL queries to 
complete its computation. In that case Event Loop run them on each iteration, 
until the Event Handler returns the status as complete.

Covers:

* `req~iterating-over-loop~1`

Needs: impl, utest, itest

### Returning JSON Result
`dsn~returning-json-result~1`

The Event Loop returns result in json format.

Covers:

* `req~returning-result~1`

Needs: impl, utest, itest

## Event Handler
The Event Handler is a framework whose logic is implemented by users. It reacts 
to the query results of  the Event Loop and generates SQL queries of the next 
state. It stores the states in BucketFS.

### Python UDF Framework 
`dsn~python-udf-framework~1`

> An Event Handler designed as Python UDF  allows users to have an easy-to-use and 
> better tooling framework to build their algorithm on top it. The user can develop 

Reviewer 2: ```suggestion
An Event Handler designed as Python UDF instead of a Lua script allows users to have an easy-to-use and 
better tooling framework to build their algorithm on top of. The user can develop 
```

the algorithm as a Python UDF script in which the framework is imported. Some 
details of this framework interface are listed below:

- The event handler is implemented as a class and states are kept as attributes 
of a class to facilitate the states operations.
- The user code should implement a method `handle_event` where the logic of the 
code is placed.
- The event handler is initiated by getting the following inputs: The results of 

> the return query,  the next state information including its parameters, 

Reviewer 2: ```suggestion
the return query, the next state information including its parameters, 
```

> context object. 

Reviewer 2: If you mean the ctx object in the udfs, not sure.
```suggestion
the ExasolDB context object. 
```

- Upon completion of the algorithm, the Event Handler calls methods itself to 
remove temporary records.
- In case of an error caught in the Event Loop, the cleanup event is called for 

> the Event Handler and the temporary records are deleted.

Author: You're right; I mentioned the keep option  in the requirements document, but overlooked it here.

Covers:

* `req~user-friendly-framework-interface~1`

Needs: impl, utest, itest

### Storing States in BucketFS 
`dsn~storing-states-in-bucketfs~1`

The framework keeps states during iterations by storing them in BucketFS.

Covers:

* `req~managing-temporary-bucketfs-files~1`

Needs: impl, utest, itest

It returns the sequence of SQL queries to execute

> A query to call the same UDF to run the action for the next state

Reviewer 2: ```suggestion
a query to call the same UDF again to run the action for the next state.
```

### Returning Queries
`dsn~returning-queries~1`

The Event Handler returns a list of SQL queries to execute and the return query 
which will be called again as the next state action so that the new  state 
transition is performed.

Covers:

* `feat~implementation-framework~1`

Needs: impl, utest, itest

### Managing Temporary BucketFS Files
`dsn~managing-temporary-bucketfs-files~1`

> The Event Handler create temporary BucketFS files that will be  kept as result. 

Author: when the execution of the framework is completed, we will no longer need these files. We might have finalization step where we cleanup them

> These temporary files are placed  in the same directory in  the BucketFS.   

Reviewer 2: ```suggestion
These temporary files are placed in the same directory in the BucketFS.   
```

Covers:

* `req~managing-temporary-bucketfs-files~1`

Needs: impl, utest, itest

### Managing Temporary BucketFS Files
`dsn~managing-temporary-tables~1`

The Event Handler uses temporary tables to store large intermediate results. 

Covers:

* `req~managing-temporary-tables~1`

Needs: impl, utest, itest

### Electing Leader
`dsn~electing-leader~1`

The leader selection consist of 3 main steps: Firstly, each UDF is assigned a 
unique id which is the concatenation of node_id and vm_id. Secondly, the list of 
UDF instances are discovered using pycos. Thirdly,  the UDF instance with the 
largest id is elected as the leader. Other instances send a confirmation message 
to the  instance with the highest id that it is the leader. If the leader got 
all the confirmation, it acknowledges them. 

Covers:

* `req~electing-leader~1`

Needs: impl, utest, itest

### Collective Operation
`dsn~collective-operation~1`

> The framework uses collective operation approach, where tasks are simultaneously 

Reviewer 2: ```suggestion
The framework uses a collective operation approach, where tasks are simultaneously 
```

> run on multiple UDF instances to achieve parallelism.

Reviewer 2: Maybe expand on this a bit? (If you know more already.)

Covers:

* `req~collective-operation~1`

Needs: impl, utest, itest

### Error Handling
In case of any errors in the execution, temporarily created files and tables 
are cleaned. 

#### Cleanup Temporary Tables
`dsn~cleanup-temporary-tables~1`

> Temporary tables are removed in case of an error that occur during the execution 

Reviewer 2: ```suggestion
Temporary tables are removed in case of an error occuring during the execution 
```

> of the framework with the keeping option  not selected.

Reviewer 2: ```suggestion
of the framework with the keeping option not selected.
```

    - To this type of execution, the model of an event handler fits best.
    - This means the Lua Script is our event loop.

## System Design and Architecture
 
![System Design Diagram](./images/system_design_diagram.png "System Design Diagram")

The designed event-driven framework consists of 3 
main components: (1) Even Loop that handles only the state transitions, (2) 
Event Handler Framework that defines a state machine and provide a framework to 
the user code, (3) Event Handler, where the user implements their algorithm. 
The Event Loop component is proposed to be implemented in Lua, because a Lua 
script is the only way to run the dynamic SQL queries in the same transactions. 
On the other hand, the Event Handler is implemented in Python script, since 
Python simplifies the development of data analysis methods by offering a wide 
variety of data processing tools.

Below you can find the code snippet that simply indicates the Python framework 
interface. Please note that the `return_query` is wrapped into the next call to 
event handler UDF.

```python
class Result:
  query_list: List[str]
  return_query: Optional[str]

> def handle_event(row_iterator:RowIterable)->Result

Reviewer: ```suggestion
class Context:
    """The Context is part of the Python framework and provides additional functions to the user code.
    
    def getTemporaryTableManager():TemporaryTableManager
    def getTemporaryBucketFSFileManager():TemporaryBucketFSFileManager
    

def handle_event(row_iterator:RowIterable,context:Context)->Result
```

main components: (1) Even Loop that handles only the state transitions, (2) 
Event Handler Framework that defines a state machine and provide a framework to 
the user code, (3) Event Handler, where the user implements their algorithm. 
The Event Loop component is proposed to be implemented in Lua, because a Lua 
script is the only way to run the dynamic SQL queries in the same transactions. 
On the other hand, the Event Handler is implemented in Python script, since 
Python simplifies the development of data analysis methods by offering a wide 
variety of data processing tools.

Below you can find the code snippet that simply indicates the Python framework 
interface. Please note that the `return_query` is wrapped into the next call to 
event handler UDF.

```python
class Result:
    query_list: List[str]
    return_query: Optional[str]

class Context:
    """The Context is part of the Python framework and provides additional functions to the user code."""
    
    def getTemporaryTableManager():TemporaryTableManager
    def getTemporaryBucketFSFileManager():TemporaryBucketFSFileManager
    
def handle_event(row_iterator:RowIterable,context:Context)->Result
    # user code
```

> 

Reviewer: Move to `dsn~python-udf-framework~1,` I think there is a better fit

until the Event Handler returns the status as complete.

Covers:

* `req~iterating-over-loop~1`

Needs: impl, utest, itest

### Returning JSON Result
`dsn~returning-json-result~1`

The Event Loop returns its result in json format.

Covers:

* `req~returning-result~1`

Needs: impl, utest, itest

## Event Handler
The Event Handler is a framework whose logic is implemented by the users. It reacts 
to the query results of the Event Loop and generates SQL queries of the next 
state. It stores the states in BucketFS.

### Python UDF Framework 
`dsn~python-udf-framework~1`

An Event Handler designed as Python UDF instead of a Lua script allows users to have an easy-to-use and 
better tooling framework to build their algorithm on top of. The user can develop 
the algorithm as a Python UDF script in which the framework is imported. Some 
details of this framework interface are listed below:

- The event handler is implemented as a class and states are kept as attributes 
of a class to facilitate the states operations.
- The user code should implement a method `handle_event` where the logic of the 
code is placed.

> - The event handler is initiated by getting the following inputs: The results of 
> the return query, the next state information including its parameters, 
> the ExasolDB context object. 

Reviewer: ```suggestion
- The first time the framework calls the event handler, it calls its constructor with a config __init__(config:Dict[str,Any]), the config is given to the event loop and the Python UDF in JSON format and deserialized to a Dict
- The event handler is called by getting the following inputs: The results of 
the return query, the ExasolDB context object. 
```

- Upon completion of the algorithm, the Event Handler calls methods itself to 
remove temporary records.
- In case of an error caught in the Event Loop, If the keep option is not set, 
the cleanup event is called for the Event Handler and the temporary records are 
deleted. Otherwise, they are kept and can be used for further investigations such as debugging.

Covers:

* `req~user-friendly-framework-interface~1`

Needs: impl, utest, itest

### Storing States in BucketFS 
`dsn~storing-states-in-bucketfs~1`

> The framework keeps states during iterations by storing them in BucketFS.

Reviewer: ```suggestion
The framework keeps states during iterations by storing them in BucketFS.
The states includes the event handler class of the user and the context state for managing temporary tables or bucketfs files
```

Covers:

* `req~managing-temporary-bucketfs-files~1`

Needs: impl, utest, itest

It returns the sequence of SQL queries to execute

> a query to call the same UDF again to run the action for the next state.

Reviewer: To which section does this belong to

### Returning Queries
`dsn~returning-queries~1`

The Event Handler returns a list of SQL queries to execute and the return query 
which will be called again as the next state action so that the new  state 
transition is performed.

Covers:

* `feat~implementation-framework~1`

Needs: impl, utest, itest

### Managing Temporary BucketFS Files
`dsn~managing-temporary-bucketfs-files~1`

The Event Handler create temporary BucketFS files that will be  kept as result. 
These temporary files are placed in the same directory in the BucketFS.   

Covers:

* `req~managing-temporary-bucketfs-files~1`

Needs: impl, utest, itest

> ### Managing Temporary BucketFS Files

Reviewer: ```suggestion
### Managing Temporary Tables
```

# Design

## Introduction
The goal of this library is to provide a general framework to implement complex 
data analysis algorithms with Exasol. The framework basically runs the necessary 
SQL queries for a given operation in a loop until the state of the operations 
is completed.

## Constraints / Limitations

- The most efficient way to run SQL queries is to use Lua script.  Only 
available in Lua Scrips, `pquery` allows SQL queries to be run in the same 
transactions.

- The results of `pquery` are simple handles and there is a limit of 250 open 
result set handles for Lua scripts.

## Assumptions

> - Since the number open result set handles in Lua scripts is limited, it is 

Reviewer: this is also a constrain

assumed that Lua scripts is not nested more than 250 times.

> - The result of pquery is assumed to not exceed the limit in size.

Reviewer: ```suggestion
- The result of pquery is assumed to not exceed the db memory limit in size. The DB usually allows 4 GB per query and node. The UDF can consume up to 32 GB per node, but get throttle down after the 4 GB. However, the 32 GB are shared between all queries on a node.
```

> 

Reviewer: * The state of the Event Handler is not too large.
* The (de-)serialization and the upload/download of the state to the BucketFS take less than 3 seconds.

## Design Considerations

- The main logic of the framework should be implemented in Python, which is 
widely used in data processing and analysis.

- The two main components of the framework, which are Even Loop and Even Handler, 
should be decoupled so that they can be independently updated and implemented.

- The Event Handler should be able to keep track of the status of a particular 
operation by storing states in BucketFS.

## System Design and Architecture
 
TBD 

## Event Loop
`dsn~event-loop~1`
Event Loop  processes only the state transitions by executing queries returned 
by the Event Handler. This loop provides the new state transition by calling 

> the Event Handler until the state is complete. Furthermore, the Event Loop 

Reviewer: ```suggestion
the Event Handler until the Event Handler returns the stop signal with a result or an error. Furthermore, the Event Loop 
```

is responsible for getting the parameters of the Event Handler that will be 
called in the loop and forward them to it.

Covers:

* `feat~TBD~1`

Needs: impl, utest, itest

## Event Handler
The Event Handler is a framework whose logic is implemented by users. It reacts 
to the query results of  the Event Loop and generates SQL queries of the next 
state. It stores the states in BucketFS.

> 

Reviewer: Yes, the event handler has two parts, the logic part is implemented by the user, the second part which calls the user code is implemented by us and provides certain features to the user code,  such as automatic state handling ((de-)serialization, upload, read). Managing temporary tables, which means the user code can register tables as temporary tables and the framework will remove them. It also forwards wraps the UDF context as iterable and forwards it to the user code. Furthermore, it builds the query to call the Event Handler again, from a given query which gives the result with which the Event handler will get called again.
